File: leetcode/Algorithms/94.BinaryTreeInorderTraversal/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution(object):
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        result = []
        self.helper(result, root)
        return result

    def helper(self, result, node):
        if node:
            if node.left:
                self.helper(result, node.left)
            result.append(node.val)
            if node.right:
                self.helper(result, node.right)

File: leetcode/Algorithms/94.BinaryTreeInorderTraversal/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

public class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        ArrayList<Integer> result = new ArrayList<>();
        Stack<TreeNode> stack = new Stack<>();
        TreeNode tmp;

        while (root != null || !stack.isEmpty()) {
            while (root != null) {
                stack.push(root);
                root = root.left;
            }
            tmp = stack.pop();
            result.add(tmp.val);
            root = tmp.right;
        }

        return result;
    }
}
